### DISCUSSION TITLE
Does the order of locations matter in the target platform?

### DISCUSSION THREAD

**Discussion by lppedd** [eclipse-tycho/tycho#2066] [Q&A] [answered]
This is probably more related to PDE than Tycho, but since I'm going to use a `.target` file for Tycho too, it's important I understand if the order matters.

Let's say I have this `.target` content:

```xml
<target name="target-platform-remote">
	<locations>
		<location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
			<repository location="https://download.eclipse.org/eclipse/updates/4.23/"/>
			...
		</location>
		<location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
			<repository location="https://archive.eclipse.org/birt/update-site/4.9.0/"/>
			...
		</location>
		<location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
			<repository location="https://public.dhe.ibm.com/ibmdl/export/pub/software/htp/zos/tools/aqua3.3/"/>
			...
		</location>
	</locations>
	<environment>
		<os>win32</os>
		<ws>win32</ws>
		<arch>x86_64</arch>
		<nl>en_US</nl>
	</environment>
	<targetJRE path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11"/>
</target>
```

If an artifact that satisfies a `Require-Bundle` is contained in all these remote p2 repositories (maybe with different minor versions),  from where is it resolved first?

**Comment by laeubi** [comment, marked as answer]
Order does not matter. Targetfile is just a collection of IUs.

**Reply by Bananeweizen** [reply]
That answer may technically be correct, but maybe we can just rephrase the answer to indicate why it was asked in the first place: Does the order of locations in a target file influence the order of remote repositories being queried by Tycho, or does Tycho have another built-in priority scheme for sorting multiple remote repositories?

**Reply by laeubi** [reply]
> why it was asked in the first place

The questions refers to dependencies of bundles, but the target platform itself do not resolve dependencies (the IU location does in planner mode but also here order does not matter at all). So probably the question makes wrong assumptions how things work?

> Does the order of locations in a target file influence the order of remote repositories being queried by Tycho

Of course it might be that repositories on the top are queried first, but there is no guarantee nor any difference, all repositories are first queried for their metadata, so order *might* influence how fast one gets an error if one of the repositories is not there at all, but after that there is only a collection of IU of **all** specified repositories.

> Tycho have another built-in priority scheme for sorting multiple remote repositories

Order does not matter, so consequently there is no sorting, there is even demands to parallelize different locations resolving:
- https://github.com/eclipse-tycho/tycho/issues/503

**Reply by Bananeweizen** [reply]
Great, that makes it way more explicit. I also tended to have similar wrong assumptions about how artifacts are retrieved.

**Reply by lppedd** [reply]
My question could have been worded better to make it more explicit on the intent. @Bananeweizen basically did it for me, thanks!
